README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# `scutr`: SMOTE and Cluster-Based Undersampling Technique in R

<!-- badges: start -->

<!-- badges: end -->

Imbalanced training datasets impede many popular classifiers. To balance
training data, a combination of oversampling minority classes and
undersampling majority classes is necessary. This package implements the
SCUT (SMOTE and Cluster-based Undersampling Technique) algorithm, which
uses model-based clustering and synthetic oversampling to balance
multiclass training datasets.

This implementation only works on numeric training data and works best
when there are more than two classes. For binary classification
problems, other packages may be better suited.

The original SCUT paper uses SMOTE (essentially linear interpolation
between points) for oversampling and expectation maximization
clustering, which fits a mixture of Gaussian distributions to the data.
These are the default methods in `scutr`, but random oversampling as
well as some distance-based undersampling techniques are available.

## Installation

You can install the released version of scutr from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("scutr")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("s-kganz/scutr")
```

## Example Usage

We start with an imbalanced dataset that comes with the package.

``` r
library(scutr)
data(imbalance)
imbalance <- imbalance[imbalance$class %in% c(2, 3, 19, 20), ]
imbalance$class <- as.numeric(imbalance$class)

plot(imbalance$V1, imbalance$V2, col=imbalance$class)
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

``` r
table(imbalance$class)
#> 
#>   2   3  19  20 
#>  20  30 190 200
```

Then, we apply SCUT with SMOTE oversampling and k-means clustering with
seven clusters.

``` r
scutted <- SCUT(imbalance, "class", undersample = undersample_kmeans,
                usamp_opts = list(k=7))
plot(scutted$V1, scutted$V2, col=scutted$class)
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

``` r
table(scutted$class)
#> 
#>   2   3  19  20 
#> 110 110 110 110
```

The dataset is now balanced and we have retained the distribution of the
data.


NEWS.md
# scutr 0.2.0

 * Fixed an issue where inappropriate values of `k` were passed to `smotefamily::SMOTE` (#1)
 * `undersample_*` functions now accept additional keyword arguments and pass these on to underlying functions, e.g. `stats::dist`.

# scutr 0.1.0

This is the initial version of the package. The following is available.

 * `SCUT` and `SCUT_parallel` driver functions.
 * Undersampling functions: `undersample_mclust`, `undersample_kmeans`, `undersample_mindist`, `undersample_hclust`, and `undersample_tomek`.
 * Oversampling functions: `oversample_smote`.
 * `resample_random` for both oversampling and undersampling.
 * Three imbalanced toy datasets: `wine`, `bullseye`, and `imbalance`.


tests/testthat.R
library(testthat)
library(scutr)

test_check("scutr")


tests/testthat/test-scut.R
context("scut")

scutenv <- as.environment("package:scutr")
oversamplers  <- ls(envir = scutenv, pattern = "oversample_*")
undersamplers <- ls(envir = scutenv, pattern = "undersample_*")
n_oversamplers  <- length(oversamplers)
n_undersamplers <- length(undersamplers)

# Test each sampling method in turn, using random as the other method
oversamplers <- c(oversamplers, rep("resample_random", n_undersamplers))
undersamplers <- c(undersamplers, rep("resample_random", n_oversamplers))
if (length(oversamplers) != length(undersamplers)) {
  stop("Testing setup should have equal number of under/oversamplers.")
}

for (i in 1:length(undersamplers)) {
  osamp <- oversamplers[i]
  usamp <- undersamplers[i]

  test_that(paste(osamp, usamp, "have equal class distribution"), {
    scutted <-
      SCUT(wine,
           "type",
           oversample = get(osamp),
           undersample = get(usamp))
    expect_true(strictly_balanced(scutted$type))
  })

  test_that(paste(osamp, usamp, "(parallel) have equal class distribution"),
            {
              # Skip to avoid triggering a warning on > 1 core
              skip_on_os("windows")


              scutted <- SCUT_parallel(
                wine,
                "type",
                # obey the CRAN limit
                ncores = 2,
                oversample = get(osamp),
                undersample = get(usamp)
              )

              expect_true(strictly_balanced(scutted$type))
            })
}

# if on windows, a warning should fire
if (.Platform$OS.type == "windows") {
  expect_warning(
    SCUT_parallel(wine, "type"),
    regexp="SCUT_parallel runs on one core only on Windows."
  )
}

# test custom functions
# dummy func that follows the required signature
foo <- function(data, cls, cls_col, m) {
  subset <- data[data[[cls_col]] == cls,]
  subset[rep(1, m),]
}
scutted <- SCUT(wine, "type", undersample = foo)
test_that("Custom functions can be passed", {
  expect_true(strictly_balanced(scutted$type))
})

scutted <-
  SCUT(wine,
       "type",
       undersample = undersample_kmeans,
       usamp_opts = list(k = 7))
test_that("Custom arguments are passed correctly", {
  expect_true(strictly_balanced(scutted$type))
  expect_error(SCUT(wine, "type", osamp_opts = list(foo = 7)),
               regexp = "unused argument*")
})

context("bad input")
bad <- wine
bad$x1 <- as.character(bad$x1)
test_that("Non numeric columns are not allowed", {
  expect_error(SCUT(bad, "type"),
               regexp = "*Data frame must be only numeric*")
})
test_that("Class column must be present", {
  expect_error(SCUT(bad, "foo"),
               regexp = "Column not found in data*")
})
bad$x1 <- as.numeric(bad$x1)
bad$x1[1] <- NA
test_that("NAs are not allowed", {
  expect_error(SCUT(bad, "type"),
               regexp = "Data frame cannot contain NAs.")
})
test_that("oversample_smote throws error on too large a value of k", {
    expect_error(SCUT(wine, "type", osamp_opts = list(k = 100)),
                 regexp="k must be smaller than the number of observations*")
})


tests/testthat/test-tomek.R
context("Tomek undersampling")

# make a synthetic dataset
# both "b" points at x < 0.0 become Tomek links. The one at -0.1 is removed first
# followed by the one at -0.5
x1 <- runif(10, 0.5, 1)
x2 <- c(-0.1, -0.5, -0.1)
y <- c(runif(10), c(0.6, 0.4, 0.5))
class <- c(rep(1, 12), 2)
df <- data.frame(x = c(x1, x2), y = y, class = class)

if (interactive()) plot(df$x, df$y, col = df$class)

tomret <- undersample_tomek(df, "a", "class", 10, tomek = "diff")

test_that("undersampling removes tomeks", {
  expect_true(all(tomret$x > 0))
})

tomret <- undersample_tomek(df, "a", "class", 9, tomek = "diff", force_m = T)

test_that("undersampling removes tomeks first", {
  expect_true(all(tomret$x > 0))
})


R/scut.R
#' Validate a dataset for resampling.
#'
#' This functions checks that the given column is present in the data and that all columns besides the class column are numeric.
#'
#' @param data Dataframe to validate.
#' @param cls_col Column with class information.
#'
#' @return NA
validate_dataset <- function(data, cls_col) {
  if (!(cls_col %in% names(data))) {
    stop("Column not found in data: ", cls_col)
  }
  if (sum(!unlist(lapply(data, is.numeric))) == is.numeric(data[[cls_col]])) {
    stop("Data frame must be only numeric besides the class column.")
  }
  if (any(is.na(data))) {
    stop("Data frame cannot contain NAs.")
  }
}

#' SMOTE and cluster-based undersampling technique.
#'
#' This function balances multiclass training datasets. In a dataframe with `n` classes and `m` rows, the resulting dataframe will have `m / n` rows per class. \code{\link{SCUT_parallel}()} distributes each over/undersampling task across multiple cores. Speedup usually occurs only if there are many classes using one of the slower resampling techniques (e.g. \code{\link{undersample_mclust}()}). Note that \code{\link{SCUT_parallel}()} will always run on one core on Windows.
#'
#' Custom functions can be used to perform under/oversampling (see the required signature below). Parameters represented by `...` should be passsed via `osamp_opts` or `usamp_opts` as a list.
#'
#' @param cls_col The column in `data` with class membership.
#' @param data Numeric data frame.
#' @param oversample Oversampling method. Must be a function with the signature `foo(data, cls, cls_col, m, ...)` that returns a data frame, one of the `oversample_*` functions, or \code{\link{resample_random}()}.
#' @param osamp_opts List of options passed to the oversampling function.
#' @param usamp_opts List of options passed to the undersampling function.
#' @param undersample Undersampling method. Must be a function with the signature `foo(data, cls, cls_col, m, ...)` that returns a data frame, one of the `undersample_*` functions, or \code{\link{resample_random}()}.
#' @param ncores Number of cores to use with \code{\link{SCUT_parallel}()}.
#' @return A dataframe with equal class distribution.
#' @export
#'
#' @references
#' Agrawal A, Viktor HL, Paquet E (2015). 'SCUT: Multi-class imbalanced data classification using SMOTE and cluster-based undersampling.' In _2015 7th International Joint Conference on Knowledge Discovery, Knowledge Engineering and Knowledge Management (IC3K)_, volume 01, 226-234.
#'
#' Chawla NV, Bowyer KW, Hall LO, Kegelmeyer WP (2002). 'SMOTE: Synthetic Minority Over-sampling Technique.' _Journal of Artificial Intelligence Research_, 16, 321-357. ISSN 1076-9757, \doi{10.1613/jair.953}, https://www.jair.org/index.php/jair/article/view/10302.
#'
#' @examples
#' ret <- SCUT(iris, "Species", undersample = undersample_hclust,
#'             usamp_opts = list(dist_calc="manhattan"))
#' ret2 <- SCUT(chickwts, "feed", undersample = undersample_kmeans)
#' table(ret$Species)
#' table(ret2$feed)
SCUT <- function(data, cls_col, oversample = oversample_smote,
                 undersample = undersample_mclust, osamp_opts = list(), usamp_opts = list()) {
  validate_dataset(data, cls_col)

  # target number of observations per class
  m <- round(nrow(data) / length(unique(data[[cls_col]])))
  # bulid skeleton of output
  ret <- as.data.frame(matrix(nrow = 0, ncol = ncol(data)), col.names = names(data))

  for (cls in unique(data[[cls_col]])) {
    n <- sum(data[[cls_col]] == cls)
    if (n < m) {
      d_prime <- do.call(
        oversample,
        c(
          list(
            data = data,
            cls_col = cls_col,
            cls = cls,
            m = m
          ),
          osamp_opts
        )
      )

      ret <- rbind(ret, d_prime)
    }
    else if (n > m) {
      d_prime <- do.call(
        undersample,
        c(
          list(
            data = data,
            cls_col = cls_col,
            cls = cls,
            m = m
          ),
          usamp_opts
        )
      )

      ret <- rbind(ret, d_prime)
    }
    else {
      # this class is already balanced
      ret <- rbind(ret, data[data[[cls_col]] == cls, ])
    }
  }

  rownames(ret) <- NULL
  return(ret)
}

#' @rdname SCUT
#'
#' @export
#'
#' @importFrom parallel detectCores mclapply
#'
#' @examples
#' # SCUT_parallel fires a warning if ncores > 1 on Windows and will run on
#' # one core only.
#' ret <- SCUT_parallel(wine, "type", ncores = 1, undersample = undersample_kmeans)
#' table(ret$type)
SCUT_parallel <- function(data, cls_col, ncores = detectCores() %/% 2,
                          oversample = oversample_smote,
                          undersample = undersample_mclust,
                          osamp_opts = list(), usamp_opts = list()) {
  validate_dataset(data, cls_col)

  # windows does not support mclapply, so we have to do this serially
  if (.Platform$OS.type == "windows") {
    warning("SCUT_parallel runs on one core only on Windows.")
    ncores <- 1
  }

  # vector of classes in the data
  classes <- unique(data[[cls_col]])
  # target number of observations per class
  m <- round(nrow(data) / length(classes))

  d_prime <- mclapply(classes, function(cls) {
    n <- sum(data[[cls_col]] == cls)
    if (n < m) {
      do.call(oversample,
              c(list(
                data = data,
                cls_col = cls_col,
                cls = cls,
                m = m
              ),
              osamp_opts))
    } else if (n > m) {
      do.call(undersample,
              c(list(
                data = data,
                cls_col = cls_col,
                cls = cls,
                m = m
              ),
              usamp_opts))
    } else {
      # this class is already balanced
      data[data[[cls_col]] == cls,]
    }
  }, mc.cores=ncores)

  return(do.call("rbind", d_prime))
}


R/mindist.R
#' Undersample a dataset by iteratively removing the observation with the lowest total distance to its neighbors.
#'
#' @param subset Dataset to undersample. Aside from cls.col, must be numeric.
#' @param cls.col Column containing class information.
#' @param m Desired number of observations after undersampling.
#' @param cls Unused, but kept here for compatability with do.undersample().
#' @param dist.calc Method for distance calculation. Must be usable with dist().
#'
#' @return An undersampled dataframe.
#' @export
#'
#' @importFrom stats dist
#'
#' @examples
#' setosa <- iris[iris$Species == "setosa", ]
#' nrow(setosa)
#' undersamp <- mindist.undersample(setosa, "setosa", "Species", 50)
#' nrow(undersamp)
mindist.undersample <- function(subset, cls, cls.col, m, dist.calc="euclidean"){
    # ignore the class column in the distance matrix since it isn't numeric
    col.ind <- which(names(subset) == cls.col)
    dist.mtx <- as.matrix(dist(subset[, -col.ind], method=dist.calc))

    # number of times to iterate
    to.discard <- nrow(subset) - m
    discard.inds <- c()

    if (to.discard < 0){
        stop("Cannot undersample dataset that has fewer rows than desired amount.")
    }

    for (i in 1:to.discard){
        # calculate the index of the minimum rowSum
        min.ind <- which.min(rowSums(dist.mtx))
        # add the rowname at that index to the discard pile
        discard.inds <- c(discard.inds, rownames(dist.mtx)[min.ind])
        # remove the row/col from distance matrix
        dist.mtx <- dist.mtx[-min.ind, -min.ind]
    }
    d_prime <- subset[-as.integer(discard.inds), ]
    rownames(d_prime) <- 1:nrow(d_prime)
    return(d_prime)
}


R/oversample.R
#' Oversample a dataset by SMOTE.
#'
#' @param data Dataset to be oversampled.
#' @param cls Class to be oversampled.
#' @param cls_col Column containing class information.
#' @param m Desired number of samples in the oversampled data.
#' @param k Number of neighbors used in \code{\link[smotefamily]{SMOTE}()} to generate synthetic minority instances. This value must be smaller than the number of minority instances already present for a given class. If `NA`, `min(5, n-1)` is chosen, where n is the number of instances of the minority class.
#'
#' @return The oversampled dataset.
#' @export
#'
#' @importFrom smotefamily SMOTE
#'
#' @examples
#' table(iris$Species)
#' smoted <- oversample_smote(iris, "setosa", "Species", 100)
#' nrow(smoted)
oversample_smote <- function(data, cls, cls_col, m, k = NA) {
  col_ind <- which(names(data) == cls_col)
  orig_cols <- names(data)
  n <- sum(data[[cls_col]] == cls)
  dup_size <- ceiling(m / n)
  # set the class to whether it is equal to the minority class
  data[[cls_col]] <- as.factor(data[[cls_col]] == cls)
  # smotefamily::SMOTE breaks for one-dim datasets. This adds a dummy column
  # so SMOTE can execute in that case. This does not affect how data is synthesized
  if (ncol(data) == 2) {
    data$dummy__col__ <- 0
  }
  # SMOTE uses the k nearest neighbors to generate a new observation. This
  # k cannot be higher than the number of instances of the minority class.
  if (is.na(k)) {
      k <- min(5, n-1)
  } else if (k > n-1) {
      stop(
          paste("k must be smaller than the number of observations in a class\n",
                "k =", k, "\n",
                "n =", n, "\n",
                "class =", cls)
          )
  }
  smoteret <- SMOTE(
    data[, -col_ind],
    data[, col_ind],
    dup_size = dup_size,
    K = k
  )
  # rbind the original observations and sufficient samples of the synthetic ones
  orig <- smoteret$orig_P
  target_samp <- m - nrow(orig)
  synt <- smoteret$syn_data[sample.int(nrow(smoteret$syn_data),
    size = target_samp,
    replace = target_samp > nrow(smoteret$syn_data)
  ), ]
  d_prime <- rbind(orig, synt)
  colnames(d_prime)[ncol(d_prime)] <- cls_col
  d_prime[[cls_col]] <- cls
  # remove the dummy column if necessary
  d_prime <- d_prime[, names(d_prime) != "dummy__col__"]
  # reorder the columns to be the same as the original data
  return(d_prime[, orig_cols])
}


R/undersample.R
#' Undersample a dataset by iteratively removing the observation with the lowest total distance to its neighbors of the same class.
#'
#' @param data Dataset to undersample. Aside from `cls_col`, must be numeric.
#' @param cls_col Column containing class information.
#' @param m Desired number of observations after undersampling.
#' @param cls Class to be undersampled.
#' @param ... Additional arguments passed to \code{\link[stats]{dist}()}.
#'
#' @return An undersampled dataframe.
#' @export
#'
#' @importFrom stats dist
#'
#' @examples
#' setosa <- iris[iris$Species == "setosa", ]
#' nrow(setosa)
#' undersamp <- undersample_mindist(setosa, "setosa", "Species", 50)
#' nrow(undersamp)
undersample_mindist <- function(data, cls, cls_col, m, ...) {
  # select the class to be undersampled
  col_ind <- which(names(data) == cls_col)
  subset <- data[data[[cls_col]] == cls, ]
  dist_mtx <- as.matrix(dist(subset[, -col_ind], ...))

  # number of times to iterate
  to_discard <- nrow(subset) - m
  discard_inds <- c()

  if (to_discard < 0) {
    stop("Cannot undersample dataset that has fewer rows than desired amount.")
  }

  for (i in 1:to_discard) {
    # calculate the index of the minimum rowSum
    min_ind <- which.min(rowSums(dist_mtx))
    # add the rowname at that index to the discard pile
    discard_inds <- c(discard_inds, min_ind)
    # eliminate that row/col from affecting further calculation
    dist_mtx[, min_ind] <- 0
    dist_mtx[min_ind, ] <- max(dist_mtx)
  }
  d_prime <- subset[-as.integer(discard_inds), ]
  return(d_prime)
}

#' Stratified index sample of different values in a vector.
#'
#' @param vec Vector of values to sample from.
#' @param tot_sample Total number of samples.
#'
#' @return A vector of indices that can be used to select a balanced population of values from `vec`.
#' @export
#'
#' @examples
#' vec <- sample(1:5, 30, replace = TRUE)
#' table(vec)
#' sample_ind <- sample_classes(vec, 15)
#' table(vec[sample_ind])
sample_classes <- function(vec, tot_sample) {
  inds <- 1:length(vec)
  samp_per_class <- ceiling(tot_sample / length(unique(vec)))
  sample_ind <- sapply(
    unique(vec),
    function(x) {
      this_class <- inds[vec == x]
      if (length(this_class) == 1) {
        rep(this_class, samp_per_class)
      } else {
        sample(this_class,
          samp_per_class,
          replace = samp_per_class > sum(vec == x)
        )
      }
    }
  )
  # transpose the resulting dataframe and trim.
  # the transpose ensures that at least samp.per.class-1 instances
  # from each cluster are included
  return(t(sample_ind)[1:tot_sample])
}

#' Undersample a dataset by expectation-maximization clustering
#'
#' @param data Data to be undersampled.
#' @param cls Class to be undersampled.
#' @param cls_col Class column.
#' @param m Number of samples in undersampled dataset.
#' @param ... Additional arguments passed to \code{\link[mclust]{Mclust}()}
#'
#' @return The undersampled dataframe containing only instance of `cls`.
#' @export
#'
#' @importFrom mclust Mclust
#' @importFrom mclust mclustBIC
#'
#' @examples
#' setosa <- iris[iris$Species == "setosa", ]
#' nrow(setosa)
#' undersamp <- undersample_mclust(setosa, "setosa", "Species", 15)
#' nrow(undersamp)
undersample_mclust <- function(data, cls, cls_col, m, ...) {
  # select the class to be undersampled
  col_ind <- which(names(data) == cls_col)
  subset <- data[data[[cls_col]] == cls, ]
  # get cluster classification
  # class col is dropped bc it often results in only one cluster
  classif <- Mclust(subset[, -col_ind], verbose = FALSE, ...)$classification
  sample_ind <- sample_classes(classif, m)
  return(subset[sample_ind, ])
}

#' Undersample a dataset by kmeans clustering.
#'
#' @param data Dataset to be undersampled.
#' @param cls Class to be undersampled.
#' @param cls_col Column containing class information.
#' @param m Number of samples in undersampled dataset.
#' @param k Number of centers in clustering.
#' @param ... Additional arguments passed to \code{\link[stats]{kmeans}()}
#'
#' @return The undersampled dataframe containing only instances of `cls`.
#' @export
#'
#' @importFrom stats kmeans
#'
#' @examples
#' table(iris$Species)
#' undersamp <- undersample_kmeans(iris, "setosa", "Species", 15)
#' nrow(undersamp)
undersample_kmeans <- function(data, cls, cls_col, m, k = 5, ...) {
  # select the class to be undersampled
  col_ind <- which(names(data) == cls_col)
  subset <- data[data[[cls_col]] == cls, ]
  # get cluster classification
  # class col is dropped bc it often results in only one cluster
  classif <- kmeans(subset[, -col_ind], centers = k, ...)$cluster
  sample_ind <- sample_classes(classif, m)
  return(subset[sample_ind, ])
}

#' Undersample a dataset by hierarchical clustering.
#'
#' @param data Dataset to be undersampled.
#' @param cls Majority class that will be undersampled.
#' @param cls_col Column in data containing class memberships.
#' @param m Number of samples in undersampled dataset.
#' @param k Number of clusters to derive from clustering.
#' @param h Height at which to cut the clustering tree. `k` must be `NA` for this to be used.
#' @param ... Additional arguments passed to \code{\link[stats]{dist}()}.
#'
#' @return Undersampled dataframe containing only `cls`.
#' @export
#'
#' @importFrom stats cutree
#' @importFrom stats hclust
#'
#' @examples
#' table(iris$Species)
#' undersamp <- undersample_hclust(iris, "setosa", "Species", 15)
#' nrow(undersamp)
undersample_hclust <- function(data, cls, cls_col, m, k = 5, h = NA, ...) {
  # select the desired class
  col_ind <- which(names(data) == cls_col)
  subset <- data[data[[cls_col]] == cls, ]
  # perform hierarchical clustering
  d <- dist(subset, ...)
  tree <- hclust(d)
  classif <- cutree(tree, k = k, h = h)
  sample_ind <- sample_classes(classif, m)
  return(subset[sample_ind, ])
}

#' Undersample a dataset by removing Tomek links.
#'
#' A Tomek link is a minority instance and majority instance that are each other's nearest neighbor. This function removes sufficient Tomek links that are an instance of cls to yield m instances of cls. If desired, samples are randomly discarded to yield m rows if insufficient Tomek links are in the data.
#'
#' @param data Dataset to be undersampled.
#' @param cls Majority class to be undersampled.
#' @param cls_col Column in data containing class memberships.
#' @param m Desired number of samples in undersampled dataset.
#' @param tomek Definition used to determine if a point is considered a minority in the Tomek link definition.
#'  - `minor`: Minor classes are all those with fewer than `m` instances.
#'  - `diff`: Minor classes are all those that aren't `cls`.
#' @param force_m If `TRUE`, uses random undersampling to discard samples if insufficient Tomek links are present to yield `m` rows of data.
#' @param ... Additional arguments passed to \code{\link[stats]{dist}()}.
#'
#' @return Undersampled dataframe containing only `cls`.
#' @export
#'
#' @importFrom stats dist
#'
#' @examples
#' table(iris$Species)
#' undersamp <- undersample_tomek(iris, "setosa", "Species", 15, tomek = "diff", force_m = TRUE)
#' nrow(undersamp)
#' undersamp2 <- undersample_tomek(iris, "setosa", "Species", 15, tomek = "diff", force_m = FALSE)
#' nrow(undersamp2)
undersample_tomek <- function(data, cls, cls_col, m, tomek = "minor",
                              force_m = TRUE, ...) {
  cls_vec <- data[[cls_col]]
  if (tomek == "diff") {
    is_minor <- cls_vec != cls
  } else if (tomek == "minor") {
    minor_classes <- Filter(function(c) sum(cls_vec == c) < m, unique(cls_vec))
    is_minor <- cls_vec %in% minor_classes
  }
  dmtx <- as.matrix(dist(data[, -which(names(data) == cls_col)], ...))
  diag(dmtx) <- max(dmtx)

  tomeks <- c()
  to_remove <- sum(cls_vec == cls) - m
  while (length(tomeks) < to_remove) {
    # determine all the tomek links in the data
    nearest <- apply(dmtx, 1, which.min)
    new_tomeks <- Filter(function(x) {
      neighbor <- nearest[x]
      nearest[neighbor] == x && cls_vec[x] == cls &&
        !(x %in% tomeks) && is_minor[neighbor]
    }, nearest)
    new_tomeks <- new_tomeks[!duplicated(new_tomeks)]
    if (length(new_tomeks) == 0) break
    tomeks <- c(tomeks, new_tomeks)
    dmtx[, new_tomeks] <- max(dmtx)
  }
  if ((length(tomeks) < to_remove) && force_m) {
    # add more indices until we get enough
    maybe_drop <- Filter(
      function(x) !(x %in% tomeks) && cls_vec[x] == cls,
      c(1:nrow(data))
    )
    tomeks <- c(tomeks, sample(maybe_drop, to_remove - length(tomeks)))
  }
  # drop the Tomek links then filter to the class of interest
  if (length(tomeks) > 0) {
    d_prime <- data[-tomeks, ]
  }
  else {
    d_prime <- data
  }

  return(d_prime[d_prime[[cls_col]] == cls, ])
}

#' Randomly resample a dataset.
#'
#' This function is used to resample a dataset by randomly removing or duplicating rows. It is usable for both oversampling and undersampling.
#'
#' @param data Dataframe to be resampled.
#' @param cls Class that should be randomly resampled.
#' @param cls_col Column containing class information.
#' @param m Desired number of samples.
#'
#' @return Resampled dataframe containing only `cls`.
#' @export
#'
#' @examples
#' set.seed(1234)
#' only2 <- resample_random(wine, 2, "type", 15)
resample_random <- function(data, cls, cls_col, m) {
  subset <- data[data[[cls_col]] == cls, ]
  if (m > nrow(subset)) {
    inds <- 1:nrow(subset)
  }
  else {
    inds <- c()
  }
  inds <- c(inds, sample.int(nrow(subset),
    size = m - length(inds),
    replace = (m - length(inds)) > nrow(subset)
  ))
  return(subset[inds, ])
}
